File: apps/Camera-Vitals/iOS/CameraVitals/Pipeline/FrameRingBuffer.swift
import Foundation

/// Fixed-capacity ring of preprocessed frames (planar RGB float, CHW per frame).
/// Thread-safe: capture thread appends, inference thread snapshots.
final class FrameRingBuffer {
    private let capacity: Int
    private let frameLen: Int
    private var storage: [[Float]]
    private var head = 0          // next write slot; when full this is also the oldest
    private var count = 0
    private let lock = NSLock()

    init(capacity: Int, frameLen: Int) {
        self.capacity = capacity
        self.frameLen = frameLen
        self.storage = Array(repeating: [Float](repeating: 0, count: frameLen), count: capacity)
    }

    var isFull: Bool {
        lock.lock(); defer { lock.unlock() }
        return count >= capacity
    }

    var filled: Int {
        lock.lock(); defer { lock.unlock() }
        return count
    }

    func append(_ frame: [Float]) {
        lock.lock(); defer { lock.unlock() }
        storage[head] = frame
        head = (head + 1) % capacity
        if count < capacity { count += 1 }
    }

    func reset() {
        lock.lock(); defer { lock.unlock() }
        count = 0
        head = 0
    }

    /// Latest `capacity` frames in temporal order (oldest → newest), flattened.
    /// Returns nil until the buffer has filled.
    func snapshot() -> [Float]? {
        lock.lock(); defer { lock.unlock() }
        guard count >= capacity else { return nil }
        var out = [Float]()
        out.reserveCapacity(capacity * frameLen)
        for i in 0..<capacity {
            out.append(contentsOf: storage[(head + i) % capacity])
        }
        return out
    }
}

File: apps/Camera-Vitals/Android/app/src/main/java/ai/zetic/demo/cameravitals/pipeline/FrameRingBuffer.kt
package ai.zetic.demo.cameravitals.pipeline

/**
 * Fixed-capacity ring of preprocessed frames (planar RGB float, CHW per frame).
 * Thread-safe: the camera thread appends, the inference thread snapshots.
 */
class FrameRingBuffer(private val capacity: Int, private val frameLen: Int) {
    private val storage = Array(capacity) { FloatArray(frameLen) }
    private var head = 0          // next write slot; when full this is also the oldest
    private var count = 0
    private val lock = Any()

    val isFull: Boolean get() = synchronized(lock) { count >= capacity }
    val filled: Int get() = synchronized(lock) { count }

    /** Copies the frame in, so the caller may reuse its buffer. */
    fun append(frame: FloatArray) = synchronized(lock) {
        System.arraycopy(frame, 0, storage[head], 0, frameLen)
        head = (head + 1) % capacity
        if (count < capacity) count++
    }

    fun reset() = synchronized(lock) {
        count = 0
        head = 0
    }

    /** Latest `capacity` frames in temporal order (oldest → newest), flattened; null until full. */
    fun snapshot(): FloatArray? = synchronized(lock) {
        if (count < capacity) return null
        val out = FloatArray(capacity * frameLen)
        for (i in 0 until capacity) {
            val idx = (head + i) % capacity
            System.arraycopy(storage[idx], 0, out, i * frameLen, frameLen)
        }
        out
    }
}
